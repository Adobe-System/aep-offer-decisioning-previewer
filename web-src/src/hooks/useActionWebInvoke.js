import { useState, useEffect } from "react";
import actionWebInvoke from "../utils";

export const useActionWebInvoke = ({
  actionName,
  headers = {},
  params = {},
}) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    // invoke Action
    actionWebInvoke(actionName, headers, params)
      .then((response) => {
        setData(response);
        setIsLoading(false);
      })
      .catch((e) => {
        setData(null);
        setIsLoading(false);
        setError(e);
      });
  }, []);
  return { data, isLoading, error };
};
