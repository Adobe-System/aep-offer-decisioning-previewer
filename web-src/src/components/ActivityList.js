import React from "react";
import PropTypes from "prop-types";
import { Picker, ProgressCircle, Item, Text } from "@adobe/react-spectrum";
import { useActionWebInvoke } from "../hooks/useActionWebInvoke";

const ActivityList = (props) => {
  let headers = {};
  // set the authorization header and org from the ims props object
  if (props.ims.token && !headers.authorization) {
    headers.authorization = `Bearer ${props.ims.token}`;
  }
  if (props.ims.org && !headers["x-gw-ims-org-id"]) {
    headers["x-gw-ims-org-id"] = props.ims.org;
  }
  const activities = useActionWebInvoke({
    actionName: "get-activities",
    headers: headers,
    params: {
      containerID: props.containerID,
    },
  });

  const onSelection = (id) => {
    const placements = activities.data._embedded.results
      .filter((activity) => activity._instance["@id"] === id)[0]
      ._instance["xdm:criteria"].map(
        (criteria) => criteria["xdm:placements"][0]
      );
    const allCollections = activities.data._embedded.results
      .filter((activity) => activity._instance["@id"] === id)[0]
      ._instance["xdm:criteria"].map(
        (criteria) => criteria["xdm:optionSelection"]["xdm:filter"] || null
      );

    const collections = [...new Set(allCollections)];

    props.onSelectionChange(id, placements, collections);
  };

  let picker = (
    <ProgressCircle
      id="activity-list-progress-circle"
      aria-label="Getting Offer Activities"
      isIndeterminate
      isHidden={!activities.isLoading}
      marginStart="size-100"
    />
  );

  if (activities.error) {
    picker = <Text>{activities.error.message}</Text>;
  }

  if (!activities.data && !activities.error && !activities.isLoading) {
    picker = <Text>You have no activities !</Text>;
  }

  if (activities.data) {
    picker = (
      <Picker
        id="activity-list-picker"
        width="100%"
        maxWidth="100%"
        label="Select Offer Activity"
        labelPosition="side"
        labelAlign="end"
        isRequired={true}
        aria-label="select an activity"
        items={activities.data._embedded.results.map((activity) => ({
          name: `${activity._instance["xdm:name"]}-(${activity._instance["xdm:status"]})`,
          id: activity._instance["@id"],
        }))}
        itemKey="id"
        onSelectionChange={onSelection}
      >
        {(item) => <Item key={item.id}>{item.name}</Item>}
      </Picker>
    );
  }

  return picker;
};

ActivityList.propTypes = {
  ims: PropTypes.any,
  containerID: PropTypes.string,
  onSelectionChange: PropTypes.func,
};

export default ActivityList;
