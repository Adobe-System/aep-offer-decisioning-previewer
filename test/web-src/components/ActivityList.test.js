import React from "react";
import { Provider, defaultTheme } from "@adobe/react-spectrum";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
jest.mock("../../../web-src/src/hooks/useActionWebInvoke.js");
import { useActionWebInvoke } from "../../../web-src/src/hooks/useActionWebInvoke";
import mock from "../../actions/mock";
import ActivityList from "../../../web-src/src/components/ActivityList";

afterEach(cleanup);
beforeEach(() => {
  jest.clearAllMocks();
});

const onSelectionChange = jest.fn(0);
const fakeProps = {
  ims: { token: "fake-token", org: "fake-org" },
  containerID: "fake-containerID",
  onSelectionChange: onSelectionChange,
};

describe("<ActivityList> calls custom hook", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { token: "fake-token", org: "fake-org" },
      containerID: "fake-containerID",
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<ActivityList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { authorization: "fake-token", "x-gw-ims-org-id": "fake-org" },
      containerID: "fake-containerID",
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<ActivityList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
describe("<ActivityList> on load", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("renders correctly on load", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<ActivityList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("has a <ProgressCircle> when loading", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<ActivityList {...fakeProps} />);
    expect(screen.getByLabelText("Getting Offer Activities")).toBeDefined();
  });
});

describe("<ActivityList> on load", () => {
  const mockResponse = {
    data: mock.data.offerActivity,
    isLoading: false,
    error: null,
  };
  it("renders correctly on data", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <Provider theme={defaultTheme} colorScheme={`light`}>
        <ActivityList {...fakeProps} />
      </Provider>
    );
    expect(screen.getByText("Select Offer Activity")).toBeInTheDocument();
    userEvent.click(screen.getByLabelText("Select Offer Activity"));
  });
});

describe("ActivityList on error", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: { message: "fake-error-message" },
  };
  it("renders correctly on error", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<ActivityList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show error message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<ActivityList {...fakeProps} />);
    expect(screen.getAllByText("fake-error-message")).toBeDefined();
  });
});

describe("<ActivityList> on data", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: null,
  };
  it("renders correctly on no data", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<ActivityList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show no data message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<ActivityList {...fakeProps} />);
    expect(screen.getAllByText("You have no activities !")).toBeDefined();
  });
});
