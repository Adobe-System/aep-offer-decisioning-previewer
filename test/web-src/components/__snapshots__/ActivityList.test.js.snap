// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivityList> calls custom hook call custom hooks with global parameters correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Activities"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="activity-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ActivityList> calls custom hook call custom hooks with global parameters correctly 2`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Activities"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="activity-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ActivityList> on data renders correctly on no data 1`] = `
<DocumentFragment>
  <span>
    You have no activities !
  </span>
</DocumentFragment>
`;

exports[`<ActivityList> on load renders correctly on load 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Activities"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="activity-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ActivityList on error renders correctly on error 1`] = `
<DocumentFragment>
  <span>
    fake-error-message
  </span>
</DocumentFragment>
`;
